File: src/nuspace/web/refs/apps/ops.py
"""The Apps op table: one wire path per op, one body per path.

The browser notifies ``<apps ref>.ops.<name>`` and the payload carries
that op's arguments and nothing else. There is no ``op`` key, because the
path already answered that question.

    op             payload
    --             --
    app.create     {name, source?}
    app.rename     {app_id, name}
    app.delete     {app_id}
    app.update     {app_id, source}
    app.restart    {app_id}

There is no ``app.select``: selection is the URL.

``target`` is the Ref each op's effect lands on. The four that change the
app list name ``Space.apps``; ``app.restart`` writes no kv at all and
only moves what this space is running, so it names the surface ref. A
Command's mutation slot is an address, and these are the honest ones.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from nuspace.web.refs.apps.interactions import AppOps
from nuspace.web.refs.common import Ops


if TYPE_CHECKING:
    import nu
    from nuspace.web.refs.apps.ref import AppsRef
    from nuspace.web.refs.apps.view import View


__all__ = ["dispatch"]


def dispatch(base: str, ref: AppsRef, view: View) -> nu.Nu:
    """Build the reactive handler for every op, composed in parallel."""
    ops = Ops(base)
    app = AppOps(view)
    kv = view.root.apps

    handlers = [
        ops.on("app.create", kv, app.create),
        ops.on("app.rename", kv, app.rename),
        ops.on("app.delete", kv, app.delete),
        ops.on("app.update", kv, app.update),
        ops.on("app.restart", ref, app.restart),
        ops.stray(ref),
    ]
    composed = handlers[0]
    for handler in handlers[1:]:
        composed = composed | handler
    return composed

File: pkgs/ts/nuspace-ui/src/refs/viewer/ops.ts
// The Viewer op table, browser side. Mirrors
// `nuspace/web/viewer/interactions.py`.
//
// One ref per op: a notify goes to `<viewer ref>.ops.<name>` and the payload
// carries that op's arguments and nothing else. There is no `op` key, because
// the path already answered that question, and the server binds one arm per
// path rather than branching on a string.
//
// `Ops` is what keeps the two sides in step: the keys are the op names the
// python ref subscribes to, and each value is exactly the field set the
// matching `nuspace.ops` call reads. Add an interaction on one side and this
// file is where the other side fails to compile.
//
// Every Cell write is here and every Plane write is in ../sidebar/ops.ts. That
// is the whole line between the two refs, and `page.select` is on this side of
// it because what the URL names is a fact about what the Viewer has open.
//
// ## Ids are minted here
//
// `section.create` carries the id of the Cell being made. That makes a create
// a pure function of its event on the server, and lets the caret aim at a
// known id rather than guess at one from a position. Use `mintId` from
// ../../app/ids; do not invent a second scheme.
//
// ## What is NOT here
//
// Typing in a text block. A text block's ref is wired straight to kv by the
// block's own program, so a keystroke is a notify on that ref and never
// touches the Viewer.
//
// Split and merge. Both are compositions of the ops below -- a split is an
// update plus a create at the next index, a merge is an update plus a delete
// -- and the canvas composes them rather than the server growing a verb for
// each editor gesture. The two halves are independent writes to different
// sections, so it does not matter which arm runs first.

import type { BlockTpl } from "./types";

/** Every op the Viewer sends, with its argument shape. */
export type Ops = {
	/** No kv write: the answer is a `set_page` plus a `set_status`. */
	"page.select": { page_id: string };
	"section.create": {
		page_id: string;
		section_id: string;
		name: string;
		tpl: BlockTpl;
		/** The Nu program. For a templated tpl this is the shipped starter,
		 *  which rides the chain onto the ref as a declared prop -- never
		 *  written here. */
		source: string;
		index: number;
	};
	/** Replaces the stored program. Nothing else about the section moves. */
	"section.update": { page_id: string; section_id: string; source: string };
	"section.delete": { page_id: string; section_id: string };
	/** Keeps the section's id, and so its kv namespace, on the new Plane. */
	"section.move": {
		page_id: string;
		section_id: string;
		to_page_id: string;
		index: number;
	};
	"section.reorder": { page_id: string; section_ids: string[] };
};

/** Send one op. Threaded down from `viewer.tsx` into the canvas. */
export type Notify = <K extends keyof Ops>(op: K, args: Ops[K]) => void;
